# Lire 

Pour les questions 01 et suivantes

* https://towardsdatascience.com/on-the-importance-of-bayesian-thinking-in-everyday-life-a74475fcceeb

# Problème 00 - Proba de sortir un 6 sachant que le dé n'est pas pipé

* Quelle est la probabilité de sortir un 6 sachant que le dé n'est pas truqué ?
* $ P(H | D) = \frac{P(D|H) * P(H)}{P(D)} $

* P(D | H) : likelihood
  * vraissemblance. 
  * Si on suppose que la théorie est vraie, quelle serait la proba d'obtenir les données en question ?
  * La probabilité que le dé ne soit pas truqué sachant qu’on a sorti un 6
  * = 1

* P(H) : proba de la théorie en l'état de nos connaissances. 
  * Avant qu'on ait connaissance des données de l'expérience. 
  * Le préjugé. 
  * Le jugement avant observation.
  * la probabilité de sortir un 6 avec un dé à 6 faces
  * = 1/6

* P(D) : Terme de normalisation. 
  * La probabilité d'obtenir ces données que ce soit avec cette théorie ou une théorie alternative quelconque. 
  * $P(D) = \sum P(D|H)*P(H) $ 
  * Le plus dur à déterminer. 
  * On peut s'en sortir si on veut juste comparer une théorie à une autre en faisant le rapport des deux
  * La probabilité que le dé ne soit pas truqué (c’est une hypothèse)
  * = 1

* P(H | D) : posterior probability. 
  * Proba à posteriori
  * Après avoir eu accès aux données. 
  * Quelle est la pobabilité que la théorie (l'hypothèse) explique les données ?
  * Quelle est la probabilité de sortir un 6 sachant que le dé n'est pas truqué ?
  * $ P(H | D) = P(\text{Sort un 6}|\text{Pas truqué}) = \frac{P(\text{Pas truqué} | \text{Sort un 6}) * P(\text{Sort un 6})}{P(\text{Pas truqué})} = \frac{1*\frac{1}{6}}{1} = \frac{1}{6} $


# Problème 0 - Proba que le dé ne soit pas pipé sachant qu'on vient de sortir un 6

* Quelle est la probabilité que le dé ne soit pas pipé sachant que je viens de sortir un 6
* $ P(H | D) = \frac{P(D|H) * P(H)}{P(D)} $

* P(D | H) : likelihood
  * vraissemblance. 
  * Si on suppose que la théorie est vraie, quelle serait la proba d'obtenir les données en question ?
  * La probabilité de sortir un 6 avec un dé non pipé
  * = 1/6

* P(H) : proba de la théorie en l'état de nos connaissances. 
  * Avant qu'on ait connaissance des données de l'expérience. 
  * Le préjugé. 
  * Le jugement avant observation.
  * la probabilité que le dé ne soit pas pipé. 
  * Cette dernière dépend du degré de trucage du dé
  * On peut supposer qu’elle est faible. 
  * Si on considère qu’il y a 5% de chance que le dé soit pipé, alors P(H)=0.95.

* P(D) : Terme de normalisation. 
  * La probabilité d'obtenir ces données que ce soit avec cette théorie ou une théorie alternative quelconque. 
  * $P(D) = \sum P(D|H)*P(H) $ 
  * Le plus dur à déterminer. 
  * On peut s'en sortir si on veut juste comparer une théorie à une autre en faisant le rapport des deux
  * la probabilité de sortir un 6, en utilisant la formule des probabilités totales
  * $ P(D) = P(D|H)P(H) + P(D|Hbar)P(Hbar) = 1/6 * 0.95 + 2/6 * 0.05 = $


* P(H | D) : posterior probability. 
  * Proba à posteriori
  * Après avoir eu accès aux données. 
  * Quelle est la pobabilité que la théorie (l'hypothèse) explique les données ?
  * Quelle est la probabilité que le dé ne soit pas truqué sachant qu'on vient de sortir un 6 ?
  * $ P(H | D) = P(\text{Pas pipé}|\text{Sort un 6}) = \frac{P(\text{Sort un 6} | \text{Pas pipé}) * P(\text{Pas pipé})}{P(\text{Sort un 6})} = \frac{1/6*0.95}{1/6 * 0.95 + 2/6 * 0.05} = 0.86 $


# Problème 1 - Fake coin ?
* I have two coins: one is a regular, fair coin and the other is a fake coin with heads of both sides. 
* I pick one of them at random, toss it, and get heads. 
* What is the probability that I tossed the fake coin?

$ P(H | D) = \frac{P(D|H) * P(H)}{P(D)} $

* P(H | D) : posterior probability. Proba à posteriori, après avoir eu accès aux données. Quelle est la pobabilité que la théorie (l'hypothèse) explique les données ?
* P(H) : proba de la théorie en l'état de nos connaissances. Avant qu'on ait connaissance des données de l'expérience. Le préjugé. Le jugement avant observation.
* P(D | H) : likelihood, vraissemblance. Si on suppose que la théorie est vraie, quelle serait la proba d'obtenir ces données ?
* P(D) : Terme de normalisation. La probabilité d'obtenir ces données que ce soit avec cette théorie ou une théorie alternative quelconque. 
  * $P(D) = \sum P(D|H)*P(H) $ 
  * Le plus dur à déterminer. 
  * On peut s'en sortir si on veut juste comparer une théorie à une autre en faisant le rapport des deux

Dans cet exemple

P(H)
* P(H) = 1/2 car on a pris une des 2 pieces au hasard

P(D|H)
* P(D|H) = 1 car si on a pris la pice truquée, à tous les coups on obtient Face

P(D)
* P(D) = On obtient une Face par la piece truquée OU on obtient une Face avec une piece normale
  * on a une chance sur 2 de choisir la piece truque => 1/2
  * on a alors 100% de chance de sortire F
  * on a une chance sur 2 de choisir la piece non truque => 1/2
  * on a alors 50% de chance de sortire F
  * P(D) = 1/2 * 100% + 1/2 * 50%

P(H|D) = 2/3

In [1]:
# prelude
import pandas as pd

In [2]:
def do_bayes_theorem(hypotheses, priors, likelihoods):
  df = pd.DataFrame({
      "hypothesis": hypotheses,
      "prior": priors,
      "likelihood": likelihoods,
  })
  df["posterior"] = df["prior"] * df["likelihood"]
  df["posterior"] /= df["posterior"].sum()
  return df

In [3]:
do_bayes_theorem(["fake_coin", "fair_coin"], [0.5, 0.5], [1, 0.5])

hypothesis  prior  likelihood  posterior
0  fake_coin    0.5         1.0   0.666667
1  fair_coin    0.5         0.5   0.333333

# Problème 2 : les 2 enfants

* Mr. Smith has two children. 
* At least one of them is a boy. 
* What is the probability that both children are boys?

In [4]:
# Faire la liste de toutes les hypotheses
hypotheses = ["GG", "BB", "GB", "BG"]

# Faire la liste des a priori
# Les préjugés. Les jugements avant observation.
priors = [1/4, 1/4, 1/4, 1/4]

# Faire la liste des vraissemblances
# Faut supposer que l'hypothèse est vraie
# Quelle chance alors d'obtenir les données
likelihoods = [0, 1/3, 1/3, 1/3]
do_bayes_theorem(hypotheses, priors, likelihoods)

hypothesis  prior  likelihood  posterior
0         GG   0.25    0.000000   0.000000
1         BB   0.25    0.333333   0.333333
2         GB   0.25    0.333333   0.333333
3         BG   0.25    0.333333   0.333333

La proba d'avoir BB est 1/3

# Problème 3 : COVID
* Je fais un test
* Il est négatif (selon le test je n'ai pas le COVID)
* proba to be infected :  0.0168 
* sensitivity is 97%. 
  * sensitivity = detection du virus chez les malades
  * Taux de Vrai Positif (TP)
  * => 3% de Faux Positifs (FP)
* specificity is 99.1%
  * specificity = detection de non virus ches les gens sains
  * Taux de Vrai Negatif (TN)
  * 0.9% de FN


|   TN  |  FP |  
|:-:    |:-:  |
|   FN  | TP  |

Precision :
* colonne de dtorite
* $ \text{Precision}  = \frac{TP}{TP+FP}$

Recall : 
* ligne du bas
* $ \text{Recall} = \frac{TP}{TP+FN}$

Sensitivity :
* ligne du bas
* $ \text{Sensitivity} = \frac{TP}{TP+FN}$

Specificity :
* ligne du haut
* $ \text{Specificity} = \frac{TN}{TN+FP}$




In [5]:
# Faire la liste de toutes les hypotheses
hypotheses = ["Santé", "Malade"]

# Faire la liste des a priori
# Les préjugés. Les jugements avant observation.
priors = [1-0.0168, 0.0168]

# Faire la liste des vraissemblances
# Faut supposer que l'hypothèse est vraie
# Quelle chance alors d'obtenir les données
likelihoods = [99.1/100, 3/100]
do_bayes_theorem(hypotheses, priors, likelihoods)

hypothesis   prior  likelihood  posterior
0      Santé  0.9832       0.991   0.999483
1     Malade  0.0168       0.030   0.000517

* 99,94% de chance d'être effectivement pas malade

Nouveaux à priori car je suis jeune et beau 
* on divise le prior par 2
* passe de 0.0168 à 0.0084

In [6]:
# Faire la liste de toutes les hypotheses
hypotheses = ["Santé", "Malade"]

# Faire la liste des a priori
# Les préjugés. Les jugements avant observation.
priors = [1-0.0084, 0.0084]

# Faire la liste des vraissemblances
# Faut supposer que l'hypothèse est vraie
# Quelle chance alors d'obtenir les données
likelihoods = [99.1/100, 3/100]
do_bayes_theorem(hypotheses, priors, likelihoods)

hypothesis   prior  likelihood  posterior
0      Santé  0.9916       0.991   0.999744
1     Malade  0.0084       0.030   0.000256

* On passe que à 99.97% (vs 99.94% avant)
* Le a posteriori est peu sensible au posteriori mais plus au likelihood

Qu'est qui se passe si je suis détecté positif ?

In [7]:
# Faire la liste de toutes les hypotheses
hypotheses = ["Santé", "Malade"]

# Faire la liste des a priori
# Les préjugés. Les jugements avant observation.
priors = [1-0.0168, 0.0168]

# Faire la liste des vraissemblances
# Faut supposer que l'hypothèse est vraie
# Quelle chance alors d'obtenir les données
likelihoods = [0.9/100, 97/100]
do_bayes_theorem(hypotheses, priors, likelihoods)

hypothesis   prior  likelihood  posterior
0      Santé  0.9832       0.009   0.351914
1     Malade  0.0168       0.970   0.648086

In [8]:
# Faire la liste de toutes les hypotheses
hypotheses = ["Santé", "Malade"]

# Faire la liste des a priori
# Les préjugés. Les jugements avant observation.
priors = [1-0.0084, 0.0084]

# Faire la liste des vraissemblances
# Faut supposer que l'hypothèse est vraie
# Quelle chance alors d'obtenir les données
likelihoods = [0.9/100, 97/100]
do_bayes_theorem(hypotheses, priors, likelihoods)

hypothesis   prior  likelihood  posterior
0      Santé  0.9916       0.009   0.522738
1     Malade  0.0084       0.970   0.477262

* Je suis détecté malade MAIS en fait j'ai plus de chance d'être en bonne santé
* Et ce malgré une sensitivité de 97%

# Problème 4 : La scène de crime

* A murder has been commited. 
* On the crime scene, two samples of blood were found: one of type “0” and one of type “AB”. 
* Type “0” is popular, 60% of the local population have it
* while type “AB” is rare, only 1% of people have it. 
* A suspect has been found whose blood type is “0”. 
* Does this evidence alone make them more or less likely to be the murderer?

In [9]:
# Faire la liste de toutes les hypotheses
hypotheses = ["Coupable", "Innocent"]

# Faire la liste des a priori
# Les préjugés. Les jugements avant observation.
priors = [1/2, 1/2]

# Faire la liste des vraissemblances
# Faut supposer que l'hypothèse est vraie
# Quelle chance d'obtenir alors les données
# Il est coupable. C'est lui qui amène le 0. La proba de trouver du AB c'est 1%
# Il est pas coupable. Faut trouver du AB et du O => 60%*1% ou du O et du AB (=>2)
# Je suis pas sûr de bien comprendre d'où vient le 2 dans l'article (question posée)
# Pour moi c'est juste : 60/100 * 1/100
likelihoods = [1/100, 60/100 * 1/100 + 1/100 * 60/100]
do_bayes_theorem(hypotheses, priors, likelihoods)

hypothesis  prior  likelihood  posterior
0   Coupable    0.5       0.010   0.454545
1   Innocent    0.5       0.012   0.545455